keras/src/trainers/epoch_iterator.py

```diff
@@ -75,55 +75,86 @@ def __init__(
-                    if len(buffer) == self.steps_per_execution:
-                        yield step - len(buffer) + 1, buffer
-                        buffer = []
-                except (StopIteration,):
-                    warnings.warn(
-                        "Your input ran out of data; interrupting epoch. "
-                        "Make sure that your dataset or generator can generate "
-                        "at least `steps_per_epoch * epochs` batches. "
-                        "You may need to use the `.repeat()` "
-                        "function when building your dataset.",
-                        stacklevel=2,
-                    )
-                    self._current_iterator = None
-                    self._insufficient_data = True
-            if buffer:
-                yield step - len(buffer) + 1, buffer
+                self._steps_seen += self.steps_per_execution
+                yield step, self._current_iterator
+            if self._num_batches and self._steps_seen >= self._num_batches:
+                self._current_iterator = iter(self._get_iterator())
+                self._steps_seen = 0
         else:
-            for step, data in enumerate(self._get_iterator()):
-                buffer.append(data)
-                if len(buffer) == self.steps_per_execution:
-                    yield step - len(buffer) + 1, buffer
-                    buffer = []
-            if buffer:
-                yield step - len(buffer) + 1, buffer
-            if not self._num_batches:
-                # Infer the number of batches returned by the data_adapter.
-                # Assumed static.
-                self._num_batches = step + 1
+            iterator = iter(self._get_iterator())
+            step = -self.steps_per_execution
+            while True:
+                step += self.steps_per_execution
+                self._steps_seen = step + self.steps_per_execution
+                yield step, iterator
         self.data_adapter.on_epoch_end()
 
+    def __iter__(self):
+        self._epoch_iterator = self._enumerate_epoch()
+        return self
+
+    def __next__(self):
+        return next(self._epoch_iterator)
+
+    @contextlib.contextmanager
+    def catch_stop_iteration(self):
+        """Catches errors when an iterator runs out of data."""
+        try:
+            yield
+        except StopIteration:
+            if self._num_batches is None:
+                self._num_batches = self._steps_seen
+            self._interrupted_warning()
+            self._current_iterator = None
+            self.data_adapter.on_epoch_end()
+
     @property
     def num_batches(self):
         if self.steps_per_epoch:
             return self.steps_per_epoch
         # Either copied from the data_adapter, or
         # inferred at the end of an iteration.
         return self._num_batches
+
+
+class EpochIterator(_EpochIterator):
+    def __next__(self):
+        buffer = []
+        step, iterator = super().__next__()
```

Comment:
Just need to call `next(self._epoch_iterator)`

Reply:
Indeed